[OSDOCS-4318]: Adds steps to use STS for 3rd-party Operators (#51541)
openshift/openshift-docs: modules/olm-configuring-aws-sts.adoc

// Module included in the following assemblies:
//
// * operators/admin/olm-adding-operators-to-cluster.adoc
// * storage/persistent_storage/persistent-storage-csi-aws-efs.adoc
// * storage/container_storage_interface/persistent-storage-csi-aws-efs.adoc
// * storage/container_storage_interface/osd-persistent-storage-aws-efs-csi.adoc

> // * storage/container_storage_interface/rosa-persistent-storage-aws-efs-csi.adoc

Author: These are here to sort out later

Perform this procedure after installing the an Operator that can use STS, but before installing any required drivers for the Operator. If you perform this procedure after installing a driver for an Operator, the Operator might not be able to function correctly.

[NOTE]
====
By default, `ccoctl` creates objects in the directory in which the commands are run. To create the objects in a different directory, use the `--output-dir` flag. This procedure uses `<path_to_ccoctl_output_dir>` to refer to this directory.

Some `ccoctl` commands make AWS API calls to create or modify AWS resources. You can use the `--dry-run` flag to avoid making API calls. Using this flag creates JSON files on the local file system instead. You can review and modify the JSON files and then apply them with the AWS CLI tool using the `--cli-input-json` parameters.
====

.Prerequisites

* You have an AWS cluster that is configured to use STS.

* You have the {product-title} release image that was used to install the cluster.

> * You have extracted and prepared the `ccoctl` binary during cluster installation. For more information, see "Configuring the Cloud Credential Operator utility".

Author: These two feel a little weird to me. Basically, we are asking them to remember what they did during install and make sure they have everything still in the same state. I am not aware of similar prereqs in our content but perhaps there are.

Reviewer: Yeah! I don't believe the user can do all of it if he already installed the cluster. To simplify those steps I would say that he can:
- extract the release image version from the cluster (CR `ClusterVersion`)
- extract the cloud-credentials-operator image from the release
- extract the `ccoctl`

```bash
RELEASE_IMAGE=$(oc get clusterversion version -o jsonpath='{.status.desired.image}')
CCO_IMAGE=$(oc adm release info --image-for='cloud-credential-operator')
oc image extract $CCO_IMAGE --file="/usr/bin/ccoctl" -a ${PULL_SECRET_FILE}
chmod 775 ccoctl
```

Author: ok, so basically the same prep as if we are starting from scratch on an install. Will grab that content :+1:

.Procedure

. Determine the name of the `CredentialsRequest` custom resource (CR) in the `openshift-cloud-credential-operator` namespace for the Operator:

> //How do they do that?

Author: In the next step, they need to refer to the CR in the `openshift-cloud-credential-operator` namespace, but I do not see how to get that string from the CSI or ALB procedures.

Author: ACK - sent you some followup in Slack

Reviewer 2: > The CredentialsRequests for the examples (CSI and ALB) were created by the Operator. I am not sure if it's a standard/required for all 3rd-party operators to create it (@abutcher thoughts?). If the Operator does not provide the CredentialsRequests resources, it should be created by the user manually.

In the STS setup for OpenShift operators, we extract the `CredentialsRequests` from the release image to create the required infrastructure ahead of installation. CCO is then ran in Manual mode, meaning there is no ongoing reconciliation of `CredentialsRequests` so I would expect the `CredentialsRequest` would be created by the user manually.

Author: per Slack convo, sounds like best strategy will be to just tell everyone how to make creds regardless of what an Operator may or may not provide (esp as minting becomes less relevant). in that case we just need a very clear, easy to fill-in YAML template in here :+1:

+
[source,terminal]
----
$ oc get credentialsrequest \
  -n openshift-cloud-credential-operator <custom_resource_name> \
  -o yaml > <path_to_directory_with_list_of_credentials_requests>/credrequests/<custom_resource_name>.yaml
----

//Existing CSI Operator has an example CR, but the CR extraction should create a YAML file that is correct without the user needing to mess with it, right?
////
. Create and save an EFS `CredentialsRequest` YAML file, such as shown in the following example, and then place it in the `credrequests` directory:
+
.Example
+
[source, yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: openshift-aws-efs-csi-driver
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - action:
      - elasticfilesystem:*
      effect: Allow
      resource: '*'
  secretRef:
    name: aws-efs-cloud-credentials
    namespace: openshift-cluster-csi-drivers
  serviceAccountNames:
  - aws-efs-csi-driver-operator
  - aws-efs-csi-driver-controller-sa
----

> ////

Author: I am leaning towards leaving this out - iiuc the user doesn't need to edit this, so it's simpler imo not to include it.

. Use the `ccoctl` tool to process all `CredentialsRequest` objects in the `credrequests` directory by running the following command:
+
[source, terminal]
----
$ ccoctl aws create-iam-roles \
  --name=<name> \
  --region=<aws_region> \

>   --credentials-requests-dir=<path_to_directory_with_list_of_credentials_requests>/credrequests \ --identity-provider-arn=arn:aws:iam::<aws_account_id>:oidc-provider/<name>-oidc.s3.<aws_region>.amazonaws.com

Author: That is not something we've included before with the `ccoctl` steps, but it certainly can't hurt :) Ideally, we would want to link to AWS steps for an AWS-side procedure so we aren't having to keep our content updated for any changes they might make. Do you have a source for it? If so, I can add it here and to the other `ccoctl` docs.

----
+
where:
+
--
** `<name>` is the name used to tag any cloud resources that are created for tracking.
** `<aws-region>` is the AWS region in which cloud resources will be created.
** `<path_to_ccoctl_output_dir>` is the path to the public key file that the `ccoctl aws create-key-pair` command generated.
** `<aws_account_id>` is the AWS account ID.
-- 
+
The command creates an IAM role in AWS for the Operator and a corresponding `<operator_credentialsrequest_name>-credentials.yaml` file in the `<path_to_ccoctl_output_dir>/manifests/` directory of the local file system.

//This step is not in the ALB proc or the existing STS content. Needs SME assessment.
. Create the Operator credentials and secret by running the following command:
+
[source, terminal]
----
$ oc create -f <path_to_ccoctl_output_dir>/manifests/<operator_credentialsrequest_name>-credentials.yaml

> ----

Author: This step is not in the ALB proc or the existing STS content. Needs SME assessment.

Author: ACK, will leave the next one in since that's what we have elsewhere.

$ ls <path_to_ccoctl_output_dir>/manifests/*-credentials.yaml | xargs -I{} oc apply -f {}
----

.Verification

//Not entirely sure this step makes sense in this version of the flow.
* To verify that the Operator secrets are created, list the files in the `<path_to_ccoctl_output_dir>/manifests` directory:
+
[source,terminal]
----
$ ll <path_to_ccoctl_output_dir>/manifests
----
+
.Example output:
+
[source,terminal]
----
total 24
-rw-------. 1 <user> <user> 161 Apr 13 11:42 cluster-authentication-02-config.yaml
-rw-------. 1 <user> <user> 379 Apr 13 11:59 openshift-cloud-credential-operator-cloud-credential-operator-iam-ro-creds-credentials.yaml
-rw-------. 1 <user> <user> 353 Apr 13 11:59 openshift-cluster-csi-drivers-ebs-cloud-credentials-credentials.yaml
-rw-------. 1 <user> <user> 355 Apr 13 11:59 openshift-image-registry-installer-cloud-credentials-credentials.yaml
-rw-------. 1 <user> <user> 339 Apr 13 11:59 openshift-ingress-operator-cloud-credentials-credentials.yaml
-rw-------. 1 <user> <user> 337 Apr 13 11:59 openshift-machine-api-aws-cloud-credentials-credentials.yaml
----
//Not sure what this should show for the Operator use case.

//This part should be right I think.

> You can verify that the IAM roles are created by querying AWS. For more information, refer to AWS documentation on listing IAM roles.

Author: The only part here that I feel confident makes sense in context is the last line (_You can verify that the IAM roles are created by querying AWS. For more information, refer to AWS documentation on listing IAM roles._). 

Does the `ll` of the `manifests` dir make sense in this version of the procedure? If it does, will need a new output showing an Operator example - this output is for cluster install.

Author: Ok, great - rearranged to accomplish this. Will just need some sample output for an `ll` in the Operator use case.

Some `ccoctl` commands make AWS API calls to create or modify AWS resources. You can use the `--dry-run` flag to avoid making API calls. Using this flag creates JSON files on the local file system instead. You can review and modify the JSON files and then apply them with the AWS CLI tool using the `--cli-input-json` parameters.
====

.Prerequisites

* You have an AWS cluster that is configured to use STS.

* You have installed an Operator that can use STS.

.Procedure

. Configure the `ccoctl` utility:

.. Get the CCO container image from the {product-title} release image by running the following command:
+
[source,terminal]
----
$ CCO_IMAGE=$(oc adm release info --image-for='cloud-credential-operator')
----

.. Extract the `ccoctl` binary from the CCO container image within the {product-title} release image by running the following command:
+
[source,terminal]
----
$ oc image extract $CCO_IMAGE --file="/usr/bin/ccoctl" -a ${PULL_SECRET_FILE}
----

> //Need to explain/define PULL_SECRET_FILE I think

Reviewer: Yeah, I agree. We can get the pull-secret in two ways:
- 1. extracting from the cluster
- 2. getting from the Console https://cloud.redhat.com/try

To 1, you can do this (as administrator):
```bash
$ oc get secret installation-pull-secrets -n openshift-image-registry \
  -o jsonpath="{.data.\.dockerconfigjson}" \
  | base64 -d > pull-secret.json
$ oc image extract $CCO_IMAGE --file="/usr/bin/ccoctl" -a ./pull-secret.json
$ rm ./pull-secret.json
```

.. Change the permissions to make `ccoctl` executable by running the following command:
+
[source,terminal]
----
$ chmod 775 ccoctl
----
////
//This strat may not actually make sense due to how STS is implemented and given the move away from minting.
ifdef::known-operator[]

> . Determine the name of the `CredentialsRequest` custom resource (CR) in the `openshift-cloud-credential-operator` namespace for the Operator: #How do they do that?#

Author: For now, I have mentioned it in the intro of the EFS CSI version to avoid messing with the rest of that installation content (out of scope and I am also concerned about messing it up). If this all look ok though I will raise it with Lisa to see if she can advise.

. Extract the CR to the `credrequests` directory by running the following command:
+
[source,terminal]
----
$ oc get credentialsrequest \
  -n openshift-cloud-credential-operator <custom_resource_name> \
  -o yaml > <path_to_directory_with_list_of_credentials_requests>/credrequests/<custom_resource_name>.yaml
----
endif::[]
////
//ifdef::olm-generic[]
. Create and save a `CredentialsRequest` custom resource (CR) YAML file, such as shown in the following example, and then place it in the `credrequests` directory:
+
.Example
+
[source, yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: openshift-aws-efs-csi-driver <1>
  namespace: openshift-cloud-credential-operator <2>
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries: <3>
    - action:
      - elasticfilesystem:*
      effect: Allow
      resource: '*'
  secretRef: 
    name: aws-efs-cloud-credentials <4>

>     namespace: openshift-cluster-csi-drivers <5>

Reviewer: Same for generic references

  serviceAccountNames: <6>
  - aws-efs-csi-driver-operator

>   - aws-efs-csi-driver-controller-sa

Reviewer: We need only one "sample" service account here.

----
<1> Specify the name of the `CredentialsRequest` CR. (#user discretion for Operators that don't provide one?#)
<2> Specify the `openshift-cloud-credential-operator` namespace.

> <3> #needinfo here#

Reviewer: I commented [here](https://github.com/openshift/openshift-docs/pull/51541#discussion_r992800823). You can reference the AWS Docs [IAM Policy Document format (regular format)](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_statement.html) to guide the users to build their own policies.

<4> #needinfo here#

> <5> #needinfo here#

Reviewer: Something like: The name of the namespace in which the Operator/Application which will consume the AWS credentials is deployed.

> <6> #needinfo here#

Reviewer: Something like: The serviceAccount used by the Operator/Application pod.

//endif::[]

. Use the `ccoctl` tool to process all `CredentialsRequest` objects in the `credrequests` directory by running the following command:
+
[source, terminal]
----
$ ccoctl aws create-iam-roles \
  --name=<name> \
  --region=<aws_region> \
  --credentials-requests-dir=<path_to_directory_with_list_of_credentials_requests>/credrequests \ 
  --identity-provider-arn=arn:aws:iam::<aws_account_id>:oidc-provider/<name>-oidc.s3.<aws_region>.amazonaws.com
----
+
where:
+
--
** `<name>` is the name used to tag any cloud resources that are created for tracking.
** `<aws-region>` is the AWS region in which cloud resources will be created.
** `<path_to_ccoctl_output_dir>` is the path to the public key file that the `ccoctl aws create-key-pair` command generated.
** `<aws_account_id>` is the AWS account ID.

> ** #need to add a link to AWS docs on how to get that ARN#

Reviewer: Added the comment on the snippet.

-- 
+

> The command creates an IAM role in AWS for the Operator and a corresponding `<operator_credentialsrequest_name>-credentials.yaml` file in the `<path_to_ccoctl_output_dir>/manifests/` directory of the local file system.

Reviewer: ```suggestion
The command creates an IAM Role in AWS for the Operator and a corresponding `<operator_credentialsrequest_name>-credentials.yaml` file in the `<path_to_ccoctl_output_dir>/manifests/` directory of the local file system.
```

. Optional: To verify that the Operator secret exists, list the contents of the `<path_to_ccoctl_output_dir>/manifests` directory by running the following command:
+
[source,terminal]
----
$ ll <path_to_ccoctl_output_dir>/manifests
----
+
.Example output:
+
[source,terminal]
----
-rw-------. 1 <user> <user> 337 Apr 13 11:59 openshift-machine-api-aws-cloud-credentials-credentials.yaml
----

> //todo: Need sample output for the Operator use case.

Reviewer: ```
$ ll cco-credrequests/manifests/
total 4
-rw-------. 1 <user> <user> 306 Oct 13 16:14 sample-sts-sample-echo-sts-credentials.yaml

```

// * storage/container_storage_interface/osd-persistent-storage-aws-efs-csi.adoc
// * storage/container_storage_interface/rosa-persistent-storage-aws-efs-csi.adoc

ifeval::["{context}" == "olm-adding-operators-to-a-cluster"]
:olm-generic:
endif::[]
ifeval::["{context}" == "persistent-storage-csi-aws-efs"]
:csi-efs-sts:
endif::[]
ifeval::["{context}" == "osd-persistent-storage-aws-efs-csi"]
:csi-efs-sts:
endif::[]
ifeval::["{context}" == "persistent-storage-csi-aws-efs"]
:csi-efs-sts:
endif::[]

////
:README:

The 'olm-generic' version of this procedure is for Operators that are not known to create a `CredentialsRequest` custom resource that the user can simply copy and apply.

For Operators that do create an appropriate CR, the `ifndef::olm-generic[]` steps apply. 

For Operators with additional special considerations, add an `ifeval` condition and include that content in a corresponding `ifdef` block (for example, the `ifdef::csi-efs-sts[]` statement below about drivers).
////

:_content-type: PROCEDURE
[id="olm-configuring-aws-sts_{context}"]
= Providing credentials for an Operator to use Amazon Web Services Secure Token Service

> You can use the CCO utility (`ccoctl`) to create resources for Operators on Amazon Web Services (AWS) clusters that use the AWS Secure Token Service (STS). Operators that use STS assign components IAM roles that provide short-term, limited-privilege security credentials. For more information, see "About manual mode with AWS Secure Token Service".

Reviewer: > "About manual mode with AWS Secure Token Service"

@jeana-redhat is it missing an URL on this reference?

  -o jsonpath="{.data.\.dockerconfigjson}" \
  | base64 -d > pull-secret.json
----

.. Extract the `ccoctl` binary from the CCO container image within the {product-title} release image by running the following command:
+
[source,terminal]
----
$ oc image extract $CCO_IMAGE --file="/usr/bin/ccoctl" -a ./pull-secret.json
----

.. Change the permissions to make `ccoctl` executable by running the following command:
+
[source,terminal]
----
$ chmod 775 ccoctl
----

.. Optional: Delete the pull secret file by running the following command:
+
[source,terminal]
----
$ rm ./pull-secret.json
----

ifndef::olm-generic[]
. Extract the `CredentialsRequest` custom resource (CR) that the Operator generated during installation by running the following command:
+
[source,terminal]
----
$ oc get credentialsrequest \
  -n <operator_cr_namespace> <operator_cr_name> \
  -o yaml > <path_to_directory_with_list_of_credentials_requests>/credrequests/<operator_cr_name>.yaml
----
where:
+
--
** `<operator_cr_namespace>` is the namespace in which the Operator generates the CR.
** `<operator_cr_name>` is the name of the CR the Operator generates.
--
endif::[]

ifdef::olm-generic[]

> . Create and save a `CredentialsRequest` custom resource (CR) YAML file, such as shown in the following example, and then place it in the `credrequests` directory:

Reviewer 2: This is the first place we mention the `credrequests` directory for the generic steps so I was thinking something like this to make it clear why we're putting the CR in a directory with a specific name. The next step mentions the `credrequests` directory too so I could be overthinking it :)

```suggestion
. Create and save a `CredentialsRequest` custom resource (CR) YAML file, such as shown in the following example, and then place it in a `credrequests` directory to be processed by `ccoctl` in a subsequent step:
```

<3> Specify IAM JSON policy statement entries. The values in this example are provided for formatting clarification only and must be updated for the requirements of the Operator. For more information, see AWS documentation about link:https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_statement.html[IAM JSON policy elements: Statement].
<4> Specify a name for the secret that will be used by the Operator. For easier reference, choose a name related to the name of the Operator.
<5> Specify the namespace in which the Operator that consumes the AWS credentials is deployed. For easier reference, choose a name related to the name of the Operator.
<6> Specify the name of the service account that is used by the Operator.
endif::[]

. Use the `ccoctl` tool to process all `CredentialsRequest` objects in the `credrequests` directory by running the following command:
+
[source, terminal]
----
$ ccoctl aws create-iam-roles \
  --name=<name> \
  --region=<aws_region> \
  --credentials-requests-dir=<path_to_directory_with_list_of_credentials_requests>/credrequests \ 

>   --identity-provider-arn=arn:aws:iam::<aws_account_id>:oidc-provider/<cluster_name>-oidc.s3.<aws_region>.amazonaws.com

Reviewer: I would add one step before this to retrieve the `issuerURL` from the Authentication object, instead of predicting this value.
This value can be changed/customized and a recently added new option using CloudFront will break this pattern.
wdyt @abutcher ?

Reviewer 2: This makes sense to me. For example, the `serviceAccountIssuer` within the authentication object could be grepped like,

```
$ oc get authentication cluster -o yaml | grep serviceAccountIssuer
  serviceAccountIssuer: https://abutcher-test-oidc.s3.us-east-1.amazonaws.com
```

Or the value of `.spec.serviceAccountIssuer` could be pulled out directly like,

```
$ oc get authentication cluster -o jsonpath --template='{ .spec.serviceAccountIssuer }'     
https://abutcher-test-oidc.s3.us-east-1.amazonaws.com
```

Reviewer: Thanks @abutcher .
@jeana-redhat ^ it should be a better approach to be used on the `--identity-provider-arn` option.

----
+
where:
+
--
** `<name>` is the name used to tag any cloud resources that are created for tracking.
** `<aws_region>` is the AWS region in which cloud resources will be created.
** `<path_to_ccoctl_output_dir>` is the path to the public key file that the `ccoctl aws create-key-pair` command generated.
** `<aws_account_id>`, `<cluster_name>`, and `<aws_region>` are standard elements of the Amazon Resource Name (ARN) for your cluster, provided here to illustrate the format of an ARN. You can obtain the ARN for your cluster's identity provider from the *Identity Providers* menu in the link:https://console.aws.amazon.com/iam/[AWS IAM console]. 
-- 
+
The command creates an IAM role in AWS for the Operator and a corresponding `<operator_secret_namespace>-<operator_secret_name>-credentials.yaml` file in the `<path_to_ccoctl_output_dir>/manifests/` directory of the local file system.

> . Optional: To verify that the Operator secret exists, list the contents of the `<path_to_ccoctl_output_dir>/manifests` directory by running the following command:

Reviewer: ```suggestion
. Optional: To verify that the secret manifests with cloud credentials exist, list the contents of the `<path_to_ccoctl_output_dir>/manifests` directory by running the following command:
```

Some `ccoctl` commands make AWS API calls to create or modify AWS resources. You can use the `--dry-run` flag to avoid making API calls. Using this flag creates JSON files on the local file system instead. You can review and modify the JSON files and then apply them with the AWS CLI tool using the `--cli-input-json` parameters.
====

.Prerequisites

* You have an AWS cluster that is configured to use STS.

* You have installed an Operator that can use STS.

* You have access to the OpenShift CLI (`oc`) that matches your {product-title} version.

* You have access to the cluster as a user with the `cluster-admin` role.

.Procedure

. Configure the `ccoctl` utility:

.. Obtain the CCO container image from the {product-title} release image by running the following command:
+
[source,terminal]
----
$ CCO_IMAGE=$(oc adm release info --image-for='cloud-credential-operator')
----

.. Save the contents of the cluster pull secret as a JSON file by running the following command:
+
[source,terminal]
----
$ oc get secret installation-pull-secrets -n openshift-image-registry \
  -o jsonpath="{.data.\.dockerconfigjson}" \
  | base64 -d > pull-secret.json
----

> .. Extract the `ccoctl` binary from the CCO container image within the {product-title} release image by running the following command:

Reviewer: I think it fits better:
```suggestion
.. Extract the `ccoctl` binary from the CCO container image by running the following command:
```

WDYT @abutcher ?

+
[source,terminal]
----
$ oc image extract $CCO_IMAGE --file="/usr/bin/ccoctl" -a ./pull-secret.json
----

.. Change the permissions to make `ccoctl` executable by running the following command:
+
[source,terminal]
----
$ chmod 775 ccoctl
----

.. Optional: Delete the pull secret file by running the following command:
+
[source,terminal]
----
$ rm ./pull-secret.json
----

ifndef::olm-generic[]
. Extract the `CredentialsRequest` custom resource (CR) that the Operator generated during installation by running the following command:
+
[source,terminal]
----
$ oc get credentialsrequest \
  -n <operator_cr_namespace> <operator_cr_name> \
  -o yaml > <path_to_directory_with_list_of_credentials_requests>/credrequests/<operator_cr_name>.yaml
----
where:
+
--
** `<operator_cr_namespace>` is the namespace in which the Operator generates the CR.
** `<operator_cr_name>` is the name of the CR the Operator generates.
--

> endif::[]

Reviewer: nit: reading the EFS rendered page[1], and the current steps[2] which provide the CredentialsRequests, and in [2] you described that the Operator provides the manifest, and needs to extract with those lines.

As we are[1] on the EFS page, it may be interesting to the reader to provide the Example of extracting it for EFS - replacing the values of `<operator_cr_namespace>` and `<operator_cr_name>` as described on [3].

[1]  https://51541--docspreview.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-aws-efs.html#olm-configuring-aws-sts_persistent-storage-csi-aws-efs
[2]  https://docs.openshift.com/container-platform/4.11/storage/container_storage_interface/persistent-storage-csi-aws-efs.html#efs-sts_persistent-storage-csi-aws-efs
[3] https://github.com/openshift/openshift-docs/pull/51541/files#diff-59156f4c6553b17ad532f0a9a551779aef65575c5dab64342843e69b2e4c38edR42

Author: I tried to get this to work with conditions, but I fear it will make quite a maintenance mess of things to have an EFS condition in the variable definitions down here. But I do think you're right to want to try to connect those values more concretely. I am going to add `<operator_cr_namespace>` and `<operator_cr_name>` up on [line 42](https://github.com/openshift/openshift-docs/pull/51541/files#diff-59156f4c6553b17ad532f0a9a551779aef65575c5dab64342843e69b2e4c38edR42) to better tie these together for the reader - will refresh the PR in a moment, lmk what you think of it :)

----
+
where:
+
--
** `<name>` is the name used to tag any cloud resources that are created for tracking.
** `<aws_region>` is the AWS region in which cloud resources will be created.
** `<path_to_ccoctl_output_dir>` is the path to the public key file that the `ccoctl aws create-key-pair` command generated.
** `<aws_account_id>` is your AWS account ID and `<issuer_URL>` is the `serviceAccountIssuer` value for your cluster. These are standard elements of the Amazon Resource Name (ARN) for your cluster, provided here to illustrate the format of an ARN. You can obtain the ARN for your cluster's identity provider from the *Identity Providers* menu in the link:https://console.aws.amazon.com/iam/[AWS IAM console].
--
+
The command creates an IAM role in AWS for the Operator and a corresponding `<operator_secret_namespace>-<operator_secret_name>-credentials.yaml` file in the `<path_to_ccoctl_output_dir>/manifests/` directory of the local file system.

. Optional: To verify that the manifests with cloud credentials exist, list the contents of the `<path_to_ccoctl_output_dir>/manifests` directory by running the following command:

> +

Reviewer 3: In format of "ccoctl aws create-iam-roles" command, parameter --credentials-requests-dir is added but no --output-dir=<path_to_ccoctl_output_dir>, 
but in the descriptions there are no <path_to_directory_with_list_of_credentials_requests> but <path_to_ccoctl_output_dir>.
Better to add --output-dir parameter and make these 2 places consistent.
